File: 445_Add_Two_Numbers_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
private:
    ListNode* reverseList(ListNode* curr) {
        ListNode* pre = NULL;
        ListNode* next = NULL;
        while (curr) {
            next = curr->next;
            curr->next = pre;
            pre = curr;
            curr = next;
        }
        return pre;
    }
    int checkCarry(ListNode* node) {
        if (node->val > 9) {
            node->val -= 10;
            return 1;
        }
        return 0;
    }

public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        l1 = reverseList(l1);
        l2 = reverseList(l2);        
        int carry = 0;
        ListNode* res = NULL;
        ListNode* pre = res;
        while (l1 || l2 || carry) {
            ListNode* node = new ListNode(0);
            if (l1 && l2) {
                node->val = l1->val + l2->val + carry--;
                carry = checkCarry(node);
                l1 = l1->next;
                l2 = l2->next;
            }
            else if (l1) {
                node->val = l1->val + carry--;
                carry = checkCarry(node);
                l1 = l1->next;
            }
            else if (l2) {
                node->val = l2->val + carry--;
                carry = checkCarry(node);
                l2 = l2->next;
            }
            else if (carry) node->val = carry--;
            
            if (res == NULL) {
                res = node;
                pre = res;
            }
            else {
                pre->next = node;
                pre = node;
            }
        }
        pre->next = NULL;
        return reverseList(res);
    }
};

File: 445_Add_Two_Numbers_II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getLen(self, head):
        size = 0
        while head:
            size += 1
            head = head.next
        return size
    
    def addToFront(self, head, val):
        node = ListNode(val)
        node.next = head
        return node

    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        n1, n2, res = self.getLen(l1), self.getLen(l2), None
        while n1 or n2:
            sum = 0
            if (n1 >= n2):
                sum += l1.val
                l1 = l1.next
                n1 -= 1
            if (n1 < n2):
                sum += l2.val
                l2 = l2.next
                n2 -= 1
            res = self.addToFront(res, sum)
        cur, carry, res = res, 0, None
        while cur:
            sum = cur.val + carry
            res = self.addToFront(res, sum % 10)
            carry = sum // 10
            tmp = cur
            cur = cur.next
            del tmp
        if carry:
            res = self.addToFront(res, carry % 10)
        return res
